Add support for STM32F0xx devices (using templates)

damaki opened the pull request:
This is a resubmission of #39 with the following updates:
- Eliminated the large number of file variants and replaced them with templated files (using the new template system).
- Bumped copyright years of the STM32F0xx source files to 2020.
- Fixed some issues in the new template system:
  - Fixed template config values not being used when installing linker scripts.
  - ~Fixed case where source files generated from templates do not use the correct line endings on Windows (by default the runtime installer was using CRLF line endings on Windows, but GNAT style rules require LF line endings).~ (this has now been fixed separately in commit 3d1542e)

I've also squashed all changes into a single commit.

damaki commented:
@Fabien-Chouteau I've just noticed that this PR is including two of your commits that implemented the template system. I updated and branched off `community-2020` when I rebased my changes a couple of days ago, but these commits no longer seem to be on `community-2020`. Were these commits recently removed from the branch?

Fabien-Chouteau commented:
Hello @damaki,

> Were these commits recently removed from the branch?

This is my fault, I will fix it ASAP.

Fabien-Chouteau commented:
@damaki It's done.

Fabien-Chouteau commented on arm/cortexm.py:
```diff
@@ -322,6 +323,232 @@ def __init__(self):
+            'arm/stm32f0xx/setup_pll.adb',
+            'arm/stm32f0xx/s-bbmcpa.ads.tmpl',
+            'arm/stm32f0xx/s-bbbopa.ads.tmpl',
+            'arm/stm32f0xx/stm32f0x{}/svd/i-stm32.ads'.format(sub_family_minor),
```
@damaki I am surprised to see so many variants of SVD. For the subset used in the run-time, what are differences?

damaki replied:
In general, the register definitions are the same. There are, however, several small (but important) differences the different sub-families:
* The interrupt table (`a-intnam.ads`) are slightly different. For example, IRQ 15 (Touch Sense Controller) is reserved and not defined on the stm32f0x0 sub-family, but is present and defined on the other sub-families (stm32f0x1, stm32f0x2, stm32f0x8). 
* Some register fields are reserved for some sub-families, but are defined for others. For example, the TIM2RST field is present in the APB1RSTR register on the stm32f0x1, but is a reserved field on the stm32f0x0.

Since these packages could be accessed from user applications (e.g. via `Ada.Interrupts.Names` and `Interfaces.STM32`), the generated run-time should contain the correct register definitions for the specific device that the run-time has been generated for. This should help detect (at compile-time) incompatibilities where a feature is assumed to be present (e.g. the TSC), but is not actually present on the specific target.

damaki commented:
@Fabien-Chouteau thanks for fixing those commits on the `community-2020` branch. FYI I have rebased (and force-pushed) my changes in this PR onto the updated branch.

Fabien-Chouteau commented on arm/stm32f0xx/README:
```diff
@@ -0,0 +1,149 @@
+ARM STM32F0 Runtimes
```
Sorry I didn't notice before, but we have an `stm32` subdir in `arm` so your `arm/stm32f0xx/` should be `arm/stm32/stm32f0xx/`

damaki replied:
Yeah I did see the `arm/stm32/` directory but unfortunately the existing code (e.g. `arm/stm32/setup_pll.adb`) is not compatible with the F0 series so I couldn't reuse any of it. So for that reason I decided to organise my code in a separate directory to avoid any clashes or confusion with the existing code.

I have also noticed that the code in PR #45 is also organised similarly to mine (in that PR, it's in `arm/stm32f1/`). I agree that we should organise the code STM32 variants into `arm/stm32/`, but I think that going forwards it would be better to reorganise the existing code into subdirectories. For example, move the existing code in `arm/stm32/` to `arm/stm32/stm32f4xx/` (or something like that). This would result in a consistent organisation of the different STM32 families under `arm/stm32/`.

Anyway, I'll move my code as you suggest. Any reorganisation of the STM32 variants can be done later.

JeremyGrosser replied:
I started on a port for the STM32G0xx chips and ran into the same problem. Personally, I like the way ChibiOS HAL approaches this problem- by identifying common low level peripherals and including different combinations of them for each chip family.
https://github.com/ChibiOS/ChibiOS/tree/master/os/hal/ports/STM32

jcassette replied:
Hi regarding #45 I created a separate folder for the same reasons you mention here

damaki replied:
@Fabien-Chouteau these have now been moved as requested in commit 90580ab

Fabien-Chouteau commented:
Hello @damaki, you have to rebase your changes.

damaki commented:
Thanks @Fabien-Chouteau , I've rebased my changes and resolved the conflict where we had both fixed the line endings problem in `support/files_holder.py` (I've removed my change in favour of yours).

Fabien-Chouteau commented on arm/cortexm.py:
```diff
@@ -2,6 +2,7 @@
 from support.bsp_sources.archsupport import ArchSupport
 from support.bsp_sources.target import Target
 
+import re
```
This file doesn't pass the Python checker (`pycodestyle` a.k.a `pep8`).

Here the changes you have to make:

```diff
diff --git a/arm/cortexm.py b/arm/cortexm.py
index d31ed21..865e710 100644
--- a/arm/cortexm.py
+++ b/arm/cortexm.py
@@ -4,6 +4,7 @@ from support.bsp_sources.target import Target
 
 import re
 
+
 class CortexMArch(ArchSupport):
     @property
     def name(self):
@@ -357,6 +358,7 @@ class CortexM0CommonArchSupport(ArmV6MTarget):
             'src/s-bbbosu__armv6m.adb',
             'src/s-bcpcst__pendsv.adb')
 
+
 class Stm32F0(CortexM0CommonArchSupport):
 
     # Flash memory size is determined by the "User code memory size"
@@ -372,7 +374,7 @@ class Stm32F0(CortexM0CommonArchSupport):
     # RAM size for STM32F03x is determined by the
     # device package (pin count) and user code memory size
     # (i.e. the last two characters of the device name, e.g. stm32f030f4)
-    f03x_ram_sizes = { # STM32F03xxx
+    f03x_ram_sizes = {  # STM32F03xxx
         'c4': 4,
         'c6': 4,
         'c8': 8,
@@ -456,7 +458,7 @@ class Stm32F0(CortexM0CommonArchSupport):
     @property
     def use_semihosting_io(self):
         return True
-    
+
     @property
     def loaders(self):
         return ('ROM', 'RAM')
@@ -467,14 +469,15 @@ class Stm32F0(CortexM0CommonArchSupport):
         # Determine MCU features from board name (e.g. 'stm32f071rb-hse')
         # The -hse or -hsi suffix specifies which clock source to
         # use for the runtime (either HSE or HSI)
-        m = re.match(r'.*f0([34579])([0128])([cefgkrv])([468bc])-(hsi|hse)', board)
+        m = re.match(r'.*f0([34579])([0128])([cefgkrv])([468bc])-(hsi|hse)',
+                     board)
         if m is None:
             raise RuntimeError("Unknown STM32F0 target: " + board)
-        sub_family_major      = m.group(1)
-        sub_family_minor      = m.group(2)
-        package               = m.group(3)
+        sub_family_major = m.group(1)
+        sub_family_minor = m.group(2)
+        package = m.group(3)
         user_code_memory_size = m.group(4)
-        clock_source          = m.group(5)
+        clock_source = m.group(5)
 
         self.board = board
 
@@ -482,11 +485,11 @@ class Stm32F0(CortexM0CommonArchSupport):
         if sub_family_major == '3':
             ram_size = self.f03x_ram_sizes[package + user_code_memory_size]
         elif sub_family_major in '45':
-            ram_size = 4 # All STM32F04x/STM32F05x devices have 4K RAM
+            ram_size = 4  # All STM32F04x/STM32F05x devices have 4K RAM
         elif sub_family_major == '7':
             ram_size = self.f07x_ram_sizes[package + user_code_memory_size]
         else:
-            ram_size = 32 # All STM32F09x devices have 32k RAM
+            ram_size = 32  # All STM32F09x devices have 32k RAM
 
         flash_size = self.flash_sizes[user_code_memory_size]
 
@@ -517,9 +520,12 @@ class Stm32F0(CortexM0CommonArchSupport):
             'arm/stm32/stm32f0xx/setup_pll.adb',
             'arm/stm32/stm32f0xx/s-bbmcpa.ads.tmpl',
             'arm/stm32/stm32f0xx/s-bbbopa.ads.tmpl',
-            'arm/stm32/stm32f0xx/stm32f0x{}/svd/i-stm32.ads'.format(sub_family_minor),
-            'arm/stm32/stm32f0xx/stm32f0x{}/svd/i-stm32-flash.ads'.format(sub_family_minor),
-            'arm/stm32/stm32f0xx/stm32f0x{}/svd/i-stm32-rcc.ads'.format(sub_family_minor))
+            ('arm/stm32/stm32f0xx/stm32f0x{}'
+             '/svd/i-stm32.ads').format(sub_family_minor),
+            ('arm/stm32/stm32f0xx/stm32f0x{}'
+             '/svd/i-stm32-flash.ads').format(sub_family_minor),
+            ('arm/stm32/stm32f0xx/stm32f0x{}'
+             '/svd/i-stm32-rcc.ads').format(sub_family_minor))
 
         # Configure MCU parameters based on family.
         if sub_family_major in '479':
@@ -546,7 +552,9 @@ class Stm32F0(CortexM0CommonArchSupport):
 
         # Choose interrupt names based on family
         self.add_gnarl_sources(
-            'arm/stm32/stm32f0xx/stm32f0x{}/svd/a-intnam.ads'.format(sub_family_minor))
+            ('arm/stm32/stm32f0xx/stm32f0x{}/'
+             'svd/a-intnam.ads').format(sub_family_minor))
+
 
 class CortexM1CommonArchSupport(ArmV6MTarget):
     @property
```

I will add a GitHub actions CI to check it in the future.

damaki replied:
Thanks for the patch Fabien. I've applied it in 3ee8458.

Fabien-Chouteau commented:
Thank you @damaki we will review the code and get back to you as soon as possible.

Fabien-Chouteau commented:
The PR is merged, thank you @damaki !
